### PR TITLE
Fix jsoc keyword not escaping a string for the query

### PR DESCRIPTION
**Pull request by nabobalis**
Supersedes: https://github.com/sunpy/sunpy/pull/7440/files

Bugfix for 5.0
Bugfix for 5.1
I would consider this a minor change.

### CODE REVIEW

File: sunpy/net/jsoc/jsoc.py [hunk 1]
```diff
@@ -650,27 +645,19 @@ def _make_recordset(self, series, start_time='', end_time='', wavelength='',
...
-                                 " or separate them by OR operator.")
-                raise ValueError(error_message)
-        else:
+            timestr = f"{primekey.pop(list(match)[0], '')}" if match else ''
+        if wavelength == '':
             # This is executed when wavelength has been passed through PrimeKey().
-            wavelength = '{}'.format(primekey.pop('WAVELNTH', ''))
+            wavelength = f"{primekey.pop('WAVELNTH', '')}"
+        elif not primekey.get('WAVELNTH'):
+            wavelength = f'{int(np.ceil(wavelength.to(u.AA).value))}'
```
Comment (wtbarnes) [inline sunpy/net/jsoc/jsoc.py:653]:
```suggestion
            wavelength = f'{int(np.ceil(wavelength.to_value(u.AA)))}'
```

File: sunpy/net/jsoc/jsoc.py [hunk 2]
```diff
@@ -702,18 +685,19 @@ def _make_recordset(self, series, start_time='', end_time='', wavelength='',
...
         keyword_info = c.keys(series)
         for key, value in keyword.items():
-            if key in keyword_info:
-                keys.append(f"{key}{value['operator']}{value['value']}")
-            else:
+            if key not in keyword_info:
                 raise ValueError(f"Keyword: '{key}' is not supported by series: {series}")
+            # We have to ensure that any values that are strings are quoted
+            other_value = value['value'] if str(value['value']).isnumeric() else f"'{value['value']}'"
+            keys.append(f"{key}{value['operator']}{other_value}")
```
Comment (wtbarnes) [inline sunpy/net/jsoc/jsoc.py:692]: This is the bugfix that is the actual focus of this PR right?

Comment (wtbarnes) [reply sunpy/net/jsoc/jsoc.py:692]: I'm also confused by this line. If as the comment says you are checking if this is a string or not, why not just do `if isinstance(value['value], str)`?

File: sunpy/net/jsoc/tests/test_jsoc.py [hunk 3]
```diff
@@ -371,3 +366,14 @@ def test_empty_response_fetch(client):
...
+
+
+@pytest.mark.remote_data
+def test_string_keyword_is_converted_correctly(client):
+    # Bug reported here:
+    # https://community.openastronomy.org/t/how-to-select-only-fuv-iris-level-1-data-using-jsoc/900
+    responses = client.search(
+        a.Time("2014-01-01T00:00", "2014-01-01T00:05"),
+        a.jsoc.Series("iris.lev1"), a.jsoc.Keyword("INSTRUME") == "FUV"
+        )
```
Comment (wtbarnes) [inline sunpy/net/jsoc/tests/test_jsoc.py:378]:
```suggestion
    )
```

Comment (wtbarnes) [reply sunpy/net/jsoc/tests/test_jsoc.py:378]: Am I going crazy? Shouldn't that that last `)` be indented at the same level as `responses`?

**Review by wtbarnes — APPROVED** [review]
This should really get a second review from someone who understands the JSOC client much better than I do.

File: sunpy/net/jsoc/tests/test_jsoc.py (hunk 3 above)
Comment (nabobalis) [reply sunpy/net/jsoc/tests/test_jsoc.py:378]: No you are correct.

File: sunpy/net/jsoc/jsoc.py (hunk 2 above)
Comment (nabobalis) [reply sunpy/net/jsoc/jsoc.py:692]: Err, I forget now

File: sunpy/net/jsoc/jsoc.py [hunk 4]
```diff
@@ -335,8 +336,8 @@ def request_data(self, jsoc_response, method='url', **kwargs):
 
         requests = []
         self.query_args = jsoc_response.query_args
-        supported_protocols = {'fits', 'as-is'}
-        supported_methods = {'url-tar', 'url', 'url-quick'}
+        supported_protocols = sorted({'fits', 'as-is'})
+        supported_methods = sorted({'url-tar', 'url', 'url-quick'})
```
Comment (Cadair) [inline sunpy/net/jsoc/jsoc.py:340]: If we care about these being sorted, why not just use a list? Taking a set and then converting it to a generator seems insane.

File: sunpy/net/jsoc/jsoc.py [hunk 5]
```diff
@@ -650,27 +645,19 @@ def _make_recordset(self, series, start_time='', end_time='', wavelength='',
                                  "prime-keys is not yet supported. Alternative is to "
                                  "use only one of the primekey to query for series data.")
                 raise ValueError(error_message)
-            if match:
```
Comment (Cadair) [inline sunpy/net/jsoc/jsoc.py:653]: As valid as these logic refactors are doing them in the middle of this PR which is fixing a legitimate bug, in code that most of us are not familiar with makes reviewing this PR *really* hard.

Comment (nabobalis) [reply sunpy/net/jsoc/jsoc.py:653]: Don't worry about it.

Comment (Cadair) [reply sunpy/net/jsoc/jsoc.py:653]: Well if you care about faster reviews then it might be something to think about :wink:

Comment (nabobalis) [reply sunpy/net/jsoc/jsoc.py:653]: *I won't reply to that with my actual thoughts.*

But yes, that is very true. I will endeavour to make sure my PR's are smaller in the future.

File: sunpy/net/jsoc/jsoc.py (hunk 4 above)
Comment (nabobalis) [reply sunpy/net/jsoc/jsoc.py:340]: Fixed.

File: sunpy/net/jsoc/jsoc.py (hunk 2 above)
Comment (nabobalis) [reply sunpy/net/jsoc/jsoc.py]: None of the unit tests failed for me locally when I did that, so I have no idea why I did that in the first place or the worrying one is that the unit tests don't cover this and I have broken something.

Only time and God can tell us.

**Comment by nabobalis** [comment]
Online fail is:

FAILED ../../.tox/py312-online/lib/python3.12/site-packages/sunpy/map/mapsequence.py::sunpy.map.mapsequence.MapSequence.save

Not due to this PR hopefully.

**Pull request merged**